system/Commands/Translation/LocalizationFinder.php

```diff
@@ -0,0 +1,370 @@
+    }
+
+    private function arrayCountRecursive(array $array, int $counter = 0): int
+    {
+        foreach ($array as $value) {
+            if (! is_array($value)) {
+                $counter++;
+            } else {
+                $counter = $this->arrayCountRecursive($value, $counter);
+            }
+        }
+
+        return $counter;
+    }
+
+    private function languageKeysDump(array $inputArray): string
+    {
+        return var_export($inputArray, true);
+    }
+
+    /**
+     * Show details in the console if the flag is set
+     */
+    private function writeIsVerbose(string $text = '', ?string $foreground = null, ?string $background = null): void
+    {
+        if ($this->verbose) {
+            CLI::write($text, $foreground, $background);
+        }
+    }
+
+    private function isSubDirectory(string $directory, string $rootDirectory): bool
+    {
+        return 0 === strncmp($directory, $rootDirectory, strlen($directory));
+    }
+
+    private function process(string $currentDir, string $currentLocale): void
+    {
+        $tableRows    = [];
+        $countNewKeys = 0;
+
+        $iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($currentDir));
+        $files    = iterator_to_array($iterator, true);
+        ksort($files);
+
+        [$languageFoundKeys, $countFiles] = $this->findLanguageKeysInFiles($files);
+        ksort($languageFoundKeys);
+
+        /**
+         * New translates
+         */
+        $languageDiff        = [];
+        $languageFoundGroups = array_unique(array_keys($languageFoundKeys));
+
+        foreach ($languageFoundGroups as $langFileName) {
+            $languageStoredKeys = [];
+            $languageFilePath   = $this->languagePath . DIRECTORY_SEPARATOR . $currentLocale . DIRECTORY_SEPARATOR . $langFileName . '.php';
+
+            if (is_file($languageFilePath)) {
+                /**
+                 * Load old localization
+                 */
+                $languageStoredKeys = require $languageFilePath;
+            }
+
+            $languageDiff = $this->arrayDiffRecursive($languageFoundKeys[$langFileName], $languageStoredKeys);
+            $countNewKeys += $this->arrayCountRecursive($languageDiff);
+
+            if ($this->showNew) {
+                $tableRows = array_merge($this->arrayToTableRows($langFileName, $languageDiff), $tableRows);
+            } else {
+                $newLanguageKeys = array_replace_recursive($languageFoundKeys[$langFileName], $languageStoredKeys);
+
+                /**
+                 * New translates exists
+                 */
+                if ($languageDiff !== []) {
+                    if (false === file_put_contents($languageFilePath, $this->templateFile($newLanguageKeys))) {
+                        $this->writeIsVerbose('Lang file ' . $langFileName . ' (error write).', 'red');
+                    } else {
+                        exec('composer cs-fix --quiet ' . $this->languagePath);
```

Comment:
`composer cs-fix` is not available by default.

And I don't know why, but I see the following output.
>   Command "cs-fix" is not defined.

Reply:
I asked you (forum):
`Do I still need to add php-cs-fixer after execution? The array dump has incorrect formatting.`
Answer: Yes. Run `composer cs-fix` always.

Task: The generated array has a syntax with array(array(), array()) and is not formatted. How can it be done differently?

Reply:
Sorry, It seems I did not understand your question correctly.
At least, we cannot use composer because it is optional to CI4 Zip installation.

Reply:
Use this https://github.com/thomasbachem/php-short-array-syntax-converter/blob/master/convert.php ?

Reply:
By the way, the `composer cs-fix` does not work in my local environment.
Does it work on your environment?

```console
$ php spark lang:find

CodeIgniter v4.4.1 Command Line Tool - Server Time: 2023-09-08 20:50:00 UTC+00:00


                                    
  Command "cs-fix" is not defined.  
                                    

All operations done!

$ cat app/Language/en/Errors.php
<?php

return array (
  'whoops' => 'Errors.whoops',
  'weHitASnag' => 'Errors.weHitASnag',
  'pageNotFound' => 'Errors.pageNotFound',
  'sorryCannotFind' => 'Errors.sorryCannotFind',
);
```

Reply:
https://github.com/codeigniter4/CodeIgniter4/blob/64819ee8df11940c42d016f1811b55e09a723198/composer.json#L99
I installed the project and executed composer update. The cs-fix command must be written in composer.json. My tests are working.

Yes, I understand that another way is needed for formatting. I didn't want to add unnecessary details, cs-fix looked good. I'll look at the options for a replacement.

Reply:
In "framework" package not exist "cs-fix" alias. Sorry.

Reply:
https://symfony.com/doc/current/components/var_exporter.html#exporting-serializing-variables
A good replacement. But this is a new dependency and dev requirement. In fact, the same cs-fix, but before saving. This solution is better and very fast.

Htitybz b which I find do not satisfy PSR-2 or are also executed as dependencies (class).

Now I'm stuck in.

Reply:
I ran this PR branch code, so I have `cs-fix` in `composer.json`.
```console
$ grep '"cs-fix":' composer.json 
        "cs-fix": [
        "cs-fix": "Fix the coding style",
```

Reply:
If we add https://github.com/symfony/var-exporter to the framework, we need to make consensus.
If you propose to do that, create another PR.

For the time being, why don't we complete this PR with the output as `array ()`?

Reply:
how about `exec('php ' . __DIR__ . '/../../../vendor/bin/php-cs-fixer fix ' . $this->languagePath)`?

Reply:
I can't confirm my opinion for adding `var-exporter`. Use the whole package for one command... 😞  Besides, how many people will discuss it? 1, 3, 5, 20?

If you use `array()`, it will overwrite all existing translations again: ` return array(...);` Is this what you expect?
If it's ok for personal use, then for developers with Git it's to re-run `cs-fix`

Reply:
> как насчет `exec('php ' . __DIR__ . '/../../../vendor/bin/php-cs-fixer fix ' . $this->languagePath)`?

That's what I originally wrote. But the `cs-fix` command in the `codeigniter4/CodeIgniter4` package looks better.
As a result, the `codeigniter4/framework` package does not have this alias. And you need to add configs with rules to this.

 As wrote @kenjis: `At least, we cannot use composer because it is optional to CI4 Zip installation.`

Reply:
> `return array(...);` Is this what you expect?

No. But I always run `composer cs-fix` before `git commit`.
So not a big problem to me.

That is a big limitation. But it is better than throwing this PR away.

Reply:
Then, how about after dumping the translations, output something like 'You need to run your linting tool to fix coding standards issues.'?

Reply:
@kenjis Yes, we are formatting the code. The other person may not even know about this possibility. Especially if we are talking about Zip installation or production without "dev".
Even if the PR disappears, I will leave it in the app/Commands for myself.

@paulbalandan I've been thinking about that, too. I didn't want to make the developer uncomfortable once again.

Reply:
I removed the formatting and added a warning. This does not disrupt the command work